packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts

```diff
@@ -0,0 +1,330 @@
+import {
+  ApolloServerPlugin,
+  GraphQLRequestListener,
+  GraphQLRequestContext,
+} from 'apollo-server-plugin-base';
+import { KeyValueCache, PrefixingKeyValueCache } from 'apollo-server-caching';
+import { ValueOrPromise } from 'apollo-server-env';
+import { CacheHint, CacheScope } from 'apollo-cache-control';
+
+// XXX This should use createSHA from apollo-server-core in order to work on
+// non-Node environments. I'm not sure where that should end up ---
+// apollo-server-sha as its own tiny module? apollo-server-env seems bad because
+// that would add sha.js to unnecessary places, I think?
+import { createHash } from 'crypto';
+import { GraphQLResponse } from 'apollo-server-core/dist/requestPipelineAPI';
+
+interface Options<TContext = Record<string, any>> {
+  // Underlying cache used to save results. All writes will be under keys that
+  // start with 'fqc:' and are followed by a fixed-size cryptographic hash of a
+  // JSON object with keys representing the query document, operation name,
+  // variables, and other keys derived from the sessionId and extraCacheKeyData
+  // hooks. If not provided, use the cache in the GraphQLRequestContext instead
+  // (ie, the cache passed to the ApolloServer constructor).
+  cache?: KeyValueCache;
+
+  // Define this hook if you're setting any cache hints with scope PRIVATE.
+  // This should return a session ID if the user is "logged in", or null if
+  // there is no "logged in" user.
+  //
+  // If a cachable response has any PRIVATE nodes, then:
+  // - If this hook is not defined, a warning will be logged and it will not be cached.
+  // - Else if this hook returns null, it will not be cached.
+  // - Else it will be cached under a cache key tagged with the session ID and
+  //   mode "private".
+  //
+  // If a cachable response has no PRIVATE nodes, then:
+  // - If this hook is not defined or returns null, it will be cached under a cache
+  //   key tagged with the mode "no session".
+  // - Else it will be cached under a cache key tagged with the mode
+  //   "authenticated public".
+  //
+  // When reading from the cache:
+  // - If this hook is not defined or returns null, look in the cache under a cache
+  //   key tagged with the mode "no session".
+  // - Else look in the cache under a cache key tagged with the session ID and the
+  //   mode "private". If no response is found in the cache, then look under a cache
+  //   key tagged with the mode "authenticated public".
+  //
+  // This allows the cache to provide different "public" results to anonymous
+  // users and logged in users ("no session" vs "authenticated public").
+  //
+  // A common implementation of this hook would be to look in
+  // requestContext.request.http.headers for a specific authentication header or
+  // cookie.
+  //
+  // This hook may return a promise because, for example, you might need to
+  // validate a cookie against an external service.
+  sessionId?(
```

Comment:
Should this be `sessionID`? I know the current codebase isn't very consistent, but I think we said we would follow https://github.com/airbnb/javascript#naming--Acronyms-and-Initialisms.

Reply:
I looked into this. apollo-server and apollo-client are almost 100% consistent in not using `ID` except stand-alone as the built-in GraphQL `ID`. This includes `clientReferenceId` and `dataIdFromObject` (an important apollo-client hook).  The only instance of `\BID` in our APIs in these two projects is `serviceID` in the plugin API, used only by the operation registry, which we should perhaps change to `serviceId` (keeping around `serviceID` for compat for a bit too).

Note that there's some code that tries to pass a `serviceID` to the EngineReportingAgent constructor too but I'm not sure why; it doesn't read it.